cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/AuthenticationStrategy.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol;
+
+import io.cdap.cdap.api.security.store.SecureStore;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import org.eclipse.jgit.transport.CredentialsProvider;
+
+/**
+ * Interface to provide authentication objects used by JGit for different
+ * {@link io.cdap.cdap.proto.sourcecontrol.Provider}s and {@link io.cdap.cdap.proto.sourcecontrol.AuthType}s.
+ */
+public interface AuthenticationStrategy {
+  /**
+   * Returns a credential provider for authenticating with a remote git repository.
+   *
+   * @param store       a {@link SecureStore} to fetch credentials
+   * @param config      Configuration for the remote repository.
+   * @param namespaceId The namespace information for fetching credentials from secure store.
+   * @return a Credential provider to be used with all git commands.
+   * @throws AuthenticationConfigException when there are problems creating the credential provider for the provided
```

Comment:
Wrt: https://github.com/cdapio/cdap/pull/14886#discussion_r1097848187
@chtyim renamed to `AuthenticationConfigException`. This is not a authentication failure, but a failure while creating the credential provider. For eg: token isn't present in secure store or GCP credentials are not found in the environment.

Reply:
Ideally, I would expect, for example, the usage of the SecureStore only happen when an authentication is needed, instead of during the construction of the provider time. You can return an implementation of the `CredentialsProvider` that wrap the `UsernamePasswordCredentialsProvider`, and only do the password lookup from the `SecureStore` when needed. This also making sure it is always using the latest password as being stored in the SecureStore, instead of some old copy.

You can either update in this PR, or have a separate PR to update it.

Reply:
Yes, that is the present behaviour. When a git command is to called, the method `RepositoryManager#createCommand()` will call `authStrategy#getCredentialsProvider()` and the token will be fetched from secure store. The repository manager stores the authStrategy and not the credentialsProvider, so the credentials used are always the latests ones.

Reply:
This is leaving the correctness about getting the latest credentials to the caller (meaning the provider cannot be cached), which is a bit count-intuitive to common practice. As I mentioned above, it can be modified to have a better contract in subsequence PR.